### PR TITLE
Added regularization to covariance in GMM maximization step to fix convergence issues in VQSR.

### PR DESCRIPTION
**Pull request by samuelklee**
A quick patch to help out the Variants team, which is struggling with a problematic callset.

Note that a similar regularization to the effective number per component probably should have been applied to solve the issue in https://github.com/broadinstitute/gatk/pull/6425. I'm not sure if the lack of this regularization will still lead to convergence issues, but I would hope that the fix that was implemented instead (treating vanishing components as a special case and skipping computation) suffices. As discussed there, we may also want to eventually remove the idiosyncratic finalize step.

The covariance regularization term added here is standard (c.f. e.g. https://github.com/scikit-learn/scikit-learn/blob/7e1e6d09bcc2eaeba98f7e737aac2ac782f0e5f1/sklearn/mixture/_gaussian_mixture.py#L154), but it may result in non-negligible changes to VQSLODs. As just discussed with the Variants team, we can probably use the WARP validation to convince ourselves that results are functionally equivalent.

I updated the exact-match tests without much close examination (by simply forcing IntegrationTestSpec.assertEqualTextFiles to overwrite the old expected files), so someone may want to sanity check them. There were also a few more interactions between the integration tests for different tools than I anticipated. Some tests use output generated by an upstream tool as input and break encapsulation.

### CODE REVIEW

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/GatherTranchesIntegrationTest.java [hunk 1]
```diff
@@ -19,7 +19,7 @@
     @Test
     public void testCombine2Shards() throws Exception {
         final File recal1 = new File(testDir + "snpTranches.scattered.txt");  //this is the output of VariantRecalibratorIntegrationTest.testVariantRecalibratorSNPscattered
-        final File recal2 = new File(testDir + "snpTranches.scattered.2.txt"); //this is the output of running the equivalent commandline as above outside of the integration test :-/
```
Comment (samuelklee) [inline src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/GatherTranchesIntegrationTest.java:22]: Not really sure why the original snpTranches.scattered.txt and snpTranches.scattered.2.txt test files were different. I would take this to indicate that running VariantRecalibratorIntegrationTest.testVariantRecalibratorSNPscattered in/outside of IntelliJ gave different results?

Just went ahead and changed this to be two copies of the VariantRecalibratorIntegrationTest.testVariantRecalibratorSNPscattered output. Although we could decouple the tests as well...

File: src/test/resources/org/broadinstitute/hellbender/tools/walkers/VQSR/expected.AS.alternate.tranches [hunk 2]
```diff
@@ -1,15 +1,15 @@
...
-97.00,0,2702,0.0000,3.2351,-2.1306,VQSRTrancheSNP90.00to97.00,SNP,2786,2702,0.9698
-98.00,0,2730,0.0000,3.2457,-2.4533,VQSRTrancheSNP97.00to98.00,SNP,2786,2730,0.9799
-99.00,0,2758,0.0000,3.2496,-3.2164,VQSRTrancheSNP98.00to99.00,SNP,2786,2758,0.9899
-99.30,0,2766,0.0000,3.2488,-4.0653,VQSRTrancheSNP99.00to99.30,SNP,2786,2766,0.9928
-99.40,0,2769,0.0000,3.2469,-5.1648,VQSRTrancheSNP99.30to99.40,SNP,2786,2769,0.9939
-99.50,0,2772,0.0000,3.2450,-5.9262,VQSRTrancheSNP99.40to99.50,SNP,2786,2772,0.9950
-99.60,0,2774,0.0000,3.2416,-6.5841,VQSRTrancheSNP99.50to99.60,SNP,2786,2774,0.9957
-99.80,0,2780,0.0000,3.2314,-58913.3494,VQSRTrancheSNP99.60to99.80,SNP,2786,2780,0.9978
-99.90,0,2783,0.0000,3.2295,-59210.6252,VQSRTrancheSNP99.80to99.90,SNP,2786,2783,0.9989
-100.00,0,2786,0.0000,3.2340,-Infinity,VQSRTrancheSNP99.90to100.00,SNP,2786,2786,1.0000
```
Comment (samuelklee) [inline src/test/resources/org/broadinstitute/hellbender/tools/walkers/VQSR/expected.AS.alternate.tranches:14]: Probably doesn't happen in real runs, but these ties and the corresponding ordering funkiness aren't great, IMO...

Comment (ldgauthier) [reply src/test/resources/org/broadinstitute/hellbender/tools/walkers/VQSR/expected.AS.alternate.tranches:14]: Yes, the "accessible truth sites" are the same (which may always be true? I don't remember), but I don't see what's weird about the ordering.

Comment (samuelklee) [reply src/test/resources/org/broadinstitute/hellbender/tools/walkers/VQSR/expected.AS.alternate.tranches:14]: Just the fact that the 99.95 target truth sensitivity appears last, after 100. I would guess that’s because the VQSLODs are tied at -Infinity and there’s nothing else in the comparator to indicate the natural ordering.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/GatherTranchesIntegrationTest.java (hunk 1 above)
Comment (ldgauthier) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/GatherTranchesIntegrationTest.java:22]: Yep, that's what happened.  It's possible I was just trying to document the existence of some non-determinism.

**Comment by ldgauthier** [comment]
@samuelklee I will build the docker and run the warp tests, but if we do this again I'm going to teach you to fish.  Wasn't there a VariantRecalibrator integration test that had some sort of random seed hijinx?  Hopefully your fix should mean we don't need any shenanigans anymore -- can you see if there are any other tests we can clean up?

File: src/test/resources/org/broadinstitute/hellbender/tools/walkers/VQSR/expected.AS.alternate.tranches (hunk 2 above)
Comment (ldgauthier) [reply src/test/resources/org/broadinstitute/hellbender/tools/walkers/VQSR/expected.AS.alternate.tranches:14]: Oh in the old one?  I guess I never noticed that.  I don't think that's worth fixing today.

File: src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/GatherTranchesIntegrationTest.java (hunk 1 above)
Comment (samuelklee) [reply src/test/java/org/broadinstitute/hellbender/tools/walkers/vqsr/GatherTranchesIntegrationTest.java:22]: Gotcha. Hmm, I would be surprised if running *only* the testVariantRecalibratorSNPscattered test through IntelliJ produced different results from running at the command line. However, I wouldn't be surprised if other tests were run in conjunction with testVariantRecalibratorSNPscattered, since that will change the state of the Utils.getRandomGenerator() (especially since the test order will be random as well).

In general, I think there's no benefit to using a global random generator. The production code in the new tools is more careful about this sort of thing, and I'll make the test code equally so.

**Comment by samuelklee** [comment]
Thanks for the quick review, @ldgauthier!

I don't think my fix will address any non-determinism in the integration tests. I'm inclined to just do better with the new tools---there does seem to be enough duct tape in the integration tests regarding re/setting the RNG so that the exact-match tests consistently pass.

As for learning how to run the WARP tests, I think that would indeed be pretty useful---for anyone that might have to update code for VQSR or the new tools in the future! Can we teach everyone to fish? Isn't this what CARROT is for?

**Comment by ldgauthier** [comment]
Warp exome JG tests had slightly different metrics, so they failed, but the NA12878 results look like this:

test
type	TP	FP	FN	RECALL	PRECISION
SNP	17444	47	843	0.954	0.997
INDEL	300	73	103	0.744	0.804

truth
type	TP	FP	FN	RECALL	PRECISION
SNP	17443	48	844	0.954	0.997
INDEL	304	74	99	0.754	0.804

(Over the exome) negligible change in SNPs, slight recall decrease in indels, which I would say is "in the noise" given how many truth indels are in the exome.

**Comment by samuelklee** [comment]
Thanks @ldgauthier! That canary certainly looks alive to me. Happy to merge whenever you or Variants team approves.

**Comment by ldgauthier** [comment]
I will approve after you take a look at the integration test I linked above.

**Comment by samuelklee** [comment]
Sorry, I guess I didn't see your edit pointing out that line of code.

I've indeed looked at that test and more---there's a lot of similar duct tape and inconsistent resetting of the RNG throughout the entire test suit. But since I think we can reasonably assume that there's enough duct tape to make things deterministic overall, I don't think it's worth cleaning up the duct tape just to get neater, but equally deterministic behavior. (Or perhaps can you point to instances of persisting non-determinism, e.g. random failures in Travis?)

In any case, I think it makes more sense to focus effort on making cleaner tests for the new tools, rather than make an 11th hour effort to revamp these existing tests. Do you agree?

See e.g. https://github.com/broadinstitute/gatk/issues/6112 for some related discussion.

**Comment by samuelklee** [comment]
As far as I can tell, even master passes when that line is removed. So the comments in the integration test are already misleading or out of date. Do you see differently?

**Review by ldgauthier — APPROVED** [review]
